v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -0,0 +1,455 @@
+            this = self_weakref()
+            message = mqtt.error_string(rc)
+
+            # To Do List
+            do_set_disconnect = False
+            do_stop_network_loop = False
+
+            # NOTE: It's not generally safe to use .is_connected() to determine what to do, since
+            # the value could change at any time. However, it IS safe to do so here.
+            # This handler, as well as .on_connect() above, are the only two functions that can
+            # change the value. They are both invoked on Paho's network loop, which is
+            # single-threaded. This means there cannot be overlapping invocations that would
+            # change the value, and thus that the value will not change during execution of this
+            # block.
+            if this.is_connected():
+                if this._desire_connection:
+                    logger.debug("Unexpected Disconnect: rc {} - {}".format(rc, message))
+                    do_set_disconnect = True
+                    do_stop_network_loop = True
+                else:
+                    logger.debug("Disconnect Response: rc {} - {}".format(rc, message))
+                    do_set_disconnect = True
+            else:
+                if this._connection_lock.connection_type == OP_TYPE_CONNECT:
+                    # Sometimes when Paho receives a failure response to a connect, the disconnect
+                    # handler is also called. Why? Who knows.
+                    # But we don't wish to issue spurious notifications or event loop stoppages.
+                    logger.debug(
+                        "Connect Failure Disconnect Response: rc {} - {}".format(rc, message)
+                    )
+                else:
+                    # Double disconnect. Suppress.
+                    # Sometimes Paho disconnects twice. Why? Who knows.
+                    # But we don't wish to issue spurious notifications or event loop stoppages.
+                    logger.debug("Double Disconnect Response: rc {} - {}".format(rc, message))
+
+            if do_set_disconnect:
+                # Change state and notify tasks waiting on disconnect
+                async def set_disconnected():
+                    logger.debug("Client State: DISCONNECTED")
+                    this._connected = False
+                    async with this.disconnected_cond:
+                        this.disconnected_cond.notify_all()
+
+                f = asyncio.run_coroutine_threadsafe(set_disconnected(), this._loop)
+                # Need to wait for this one to finish since we don't want to let another
+                # Paho handler invoke until we know the connection state has been set.
+                f.result()
+
+            if do_stop_network_loop:
+
+                async def stop_network_loop():
+                    logger.debug("Stopping Paho network loop due to unexpected disconnect")
+                    client.loop_stop()
```

Comment:
It may not ever be necessary to stop the loop. IIRC, this was added to our code before the `reconnect_on_failure` flag was added to Paho.

Reply:
It is necessary. You can't reconnect until you stop, and then restart the loop as I recall.